src/net_processing.cpp

```diff
@@ -1888,16 +1888,23 @@ void static ProcessOrphanTx(CConnman* connman, std::set<uint256>& orphan_work_se
 }
 
 /**
- *  A block has been processed. If this is the first time we've seen the block,
+ *  A block has been processed. If the block was failed anti-dos / mutation checks, then
+ *  call BlockChecked() to maybe punish peer. If this is the first time we've seen the block,
  *  update the node's nLastBlockTime. Otherwise erase it from mapBlockSource.
  */
-void static BlockProcessed(CNode* pfrom, std::shared_ptr<CBlock> pblock, bool new_block)
+void static BlockProcessed(CNode* pfrom, CConnman* connman, std::shared_ptr<CBlock> pblock, BlockValidationState& state, bool new_block)
 {
-    if (new_block) {
-        pfrom->nLastBlockTime = GetTime();
-    } else {
+    if (!state.IsValid()) {
```

Comment:
No. `BlockProcessed` is only called when `ProcessNewBlock()` returns early because the block failed `CheckBlock()` or `AcceptBlock()`. The logic you're talking about for promoting a peer to compact block announcement in `BlockChecked()` is only invoked when `BlockChecked()` is called after successfully connecting a valid block.

Reply:
Right, right, the call to `BlockChecked` is still in `ConnectTip` path on your branch. I still think that's counter-intuitive to have both peer punishment and compact blocks promotion in one function, so would be down to split in a follow-up PR and like you said previously rely on `UpdatedBlockTip` for the latter.